proposals/0187-introduce-filtermap.md

```diff
@@ -0,0 +1,127 @@
+# Introduce Sequence.filterMap(_:)
+
+* Proposal: [SE-0187](0187-introduce-filtermap.md)
+* Authors: [Max Moiseev](https://github.com/moiseev)
+* Review Manager: TBD
+* Status: **Awaiting implementation**
+* Swift-evolution discussion: https://lists.swift.org/pipermail/swift-evolution/Week-of-Mon-20171023/040609.html
+
+<!--
+* During the review process, add the following fields as needed:*
+* Implementation: [apple/swift#NNNNN](https://github.com/apple/swift/pull/NNNNN)
+* Decision Notes: [Rationale](https://lists.swift.org/pipermail/swift-evolution/), [Additional Commentary](https://lists.swift.org/pipermail/swift-evolution/)
+* Bugs: [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN), [SR-MMMM](https://bugs.swift.org/browse/SR-MMMM)
+
+* Previous Revision: [1](https://github.com/apple/swift-evolution/blob/...commit-ID.../proposals/NNNN-filename.md)
+* Previous Proposal: [SE-XXXX](XXXX-filename.md)
+
+-->
+
+
+## Introduction
+
+We propose to deprecate the controversial version of a `Sequence.flatMap` method
+and provide the same functionality under a different, and potentially more
+descriptive, name.
+
+## Motivation
+
+The Swift standard library currently defines 3 distinct overloads for `flatMap`:
+
+~~~swift
+Sequence.flatMap<S>(_: (Element) -> S) -> [S.Element]
+    where S : Sequence
+Optional.flatMap<U>(_: (Wrapped) -> U?) -> U?
+Sequence.flatMap<U>(_: (Element) -> U?) -> [U]
+~~~
+
+The last one, despite being useful in certain situations, can be (and often is)
+misused. Consider the following snippet:
+
+~~~swift
+struct Person {
+  var age: Int
+  var name: String
+}
+
+func getAges(people: [Person]) -> [Int] {
+  return people.flatMap { $0.age }
+}
+~~~
+
+What happens inside `getAges` is: thanks to the implicit promotion to
+`Optional`, the result of the closure gets wrapped into a `.some`, then
+immediately unwrapped by the implementation of `flatMap`, and appended to the
+result array. All this unnecessary wrapping and unwrapping can be easily avoided
+by just using `map` instead.
+
+~~~swift
+func getAges(people: [Person]) -> [Int] {
+  return people.map { $0.age }
+}
+~~~
+
+It gets even worse when we consider future code modifications, like the one
+where Swift 4 introduced a `String` conformance to the `Collection` protocol.
+The following code used to compile (due to the `flatMap` overload in question).
+
+~~~swift
+func getNames(people: [Person]) -> [String] {
+  return people.flatMap { $0.name }
+}
+~~~
+
+But it no longer does, because now there is a better overload that does not
```

Comment:
In this case, I think it does still compile because the type context forces the other overload to kick in. The point about it breaking when done over two statements stands, tho.

Reply:
Well, this is from the universe where we did not add [these overloads](https://github.com/apple/swift/blob/master/stdlib/public/core/StringRangeReplaceableCollection.swift.gyb#L401).